apps/vmq_generic_offline_msg_store/src/engines/vmq_offline_storage_engine_postgres.erl

```diff
@@ -0,0 +1,43 @@
+-module(vmq_offline_storage_engine_postgres).
+
+-export([open/1, close/1, write/4, delete/2, delete/3, read/3, find/2]).
+
+-include_lib("vmq_commons/src/vmq_types_common.hrl").
+
+-define(TABLE, "messages").
+
+% API
+open(Opts) ->
+    epgsql:connect(Opts).
+
+write(Client, SIdB, MsgRef, MsgB) ->
+    epgsql:equery(Client,
+                  "INSERT INTO " ++ ?TABLE ++ " (sid, msgref, payload) VALUES ($1, $2, $3)",
+                  [SIdB, MsgRef, MsgB]
+                 ).
+
+delete(Client, SIdB) ->
+    epgsql:equery(Client, "DELETE FROM " ++ ?TABLE ++ " WHERE sid=$1", [SIdB]).
+
+delete(Client, SIdB, MsgRef) ->
+    epgsql:equery(Client, "DELETE FROM " ++ ?TABLE ++ " WHERE sid=$1 AND msgref=$2", [SIdB, MsgRef]).
+
+read(Client, SIdB, MsgRef) ->
+    case epgsql:equery(Client, "SELECT payload FROM " ++ ?TABLE ++ " WHERE sid=$1 AND msgref=$2", [SIdB, MsgRef]) of
+        {ok, _, [{BinaryMsg}]} -> {ok, binary_to_term(BinaryMsg)};
+        E -> E
+    end.
+
+find(Client, SIdB) ->
```

Comment:
This may cause high latency in sending ConnAck

Reply:
For init_offline_queue which fetches all the messages for the client by calling this fn find is called asynchronously. So it won't affect connack but on adding the session we delete the offline messages if any using delete fn here, since this is a synchronous blocking call, it may happen that due to increase in response time from DB connack will be affected. We can discuss if we also need to make the delete call async.